Initial Best Practices for e2e and functional tests (#504)
integr8ly/integreatly-operator: test/BEST_PRACTICES.md

# Test Best Practices

This is a collection of best practices to write high quality tests. The best practices

> can be referenced in review comments to speed up the review process.

Reviewer: Suggested wording:
'...These best practices should be referenced in the comments that you provide when you perform code reviews of tests...'

Author: @ldimaggi if you want to make the suggestion even more effectively you can use the github suggestions so that I can accept them with one click, less work for me :)

- [Code Style](#code-style)
- [Test Case Traceability](#test-case-traceability)
- [Don't Sleep](#dont-sleep)
- [Independent](#independent)
- [Secrets](#secrets)
- [Logging](#logging)

> ## Code Style

Reviewer: Suggested - change this to "Coding Style"

The code style should be consistent with the style used in the integreatly-operator codebase [TODO: link to golang practices used by engineering]

Also, in case of doubt, the Golang standard best practices can be used:

- [Go Code Review Comments](https://github.com/golang/go/wiki/CodeReviewComments)
- [Effective Go](https://golang.org/doc/effective_go.html) documents.

## Test Case Traceability

In order to compute test coverage and test automation progress across multiple test suites all
automated tests must have an ID and must be traced back to the [integreatly-test-cases](https://gitlab.cee.redhat.com/integreatly-qe/integreatly-test-cases)
repo following the [How to automate a test case and link it back](https://gitlab.cee.redhat.com/integreatly-qe/integreatly-test-cases#how-to-automate-a-test-case-and-link-it-back)
tutorial.

## Don't Sleep

> When waiting for a resource to become available or an operation to complete do not use fixed wait times and/or sleep statements,

Reviewer: Very samll point - remove the NEWLINE.   ;-)

Author: I removed the newlines from all paragraphs, is that right

# Test Best Practices

This is a collection of best practices to write high quality tests. The best practices
can be referenced in review comments to speed up the review process.

- [Go Style](#go-style)
- [Test Case Traceability](#test-case-traceability)
- [Don't Sleep](#dont-sleep)
- [Independent](#independent)
- [Secrets](#secrets)
- [Logging](#logging)

## Go Style

> We don't want to redefine best practices for go, therefore we should reuse the existing ones

Reviewer: The reason here is that we want these best practices to focus on test design and implementation. The same coding standards used throughout the product team should also be followed for test code.

Author: The problem is that the project is completely missing a Contribution or Best Practices guide, so I don't know which shared practices eng are using

Author: @ldimaggi tried to reword it wdyt?

Reviewer 2: ```suggestion
Test code style to be consistent with that of integreatly-operator codebase [TODO: link to golang practices used by engineering]
```
perhaps something like the above?
tbh if there are no shared practices to link to, I would question the value of this section in the doc though.

Author: Try to reworded again, wdty?

Author: ping @damienomurchu

Reviewer: Looks good. Were you able to locate the Eng Golang practices doc?

Reviewer 2: I wasn't when I looked, but lets not let this block this PR. 
@b1zzu make the above change with the TODO, and we can update this doc later after its merged?

described in the [Go Code Review Comments](https://github.com/golang/go/wiki/CodeReviewComments)

> and in the[Effective Go](https://golang.org/doc/effective_go.html) documents.

Reviewer 2: ```suggestion
and in the [Effective Go](https://golang.org/doc/effective_go.html) documents.
```

## Test Case Traceability

In order to compute test coverage and test automation progress across multiple test suites all
automated tests must have an ID and must be traced back to the [integreatly-test-cases](https://gitlab.cee.redhat.com/integreatly-qe/integreatly-test-cases)
repo following the [How to automate a test case and link it back](https://gitlab.cee.redhat.com/integreatly-qe/integreatly-test-cases#how-to-automate-a-test-case-and-link-it-back)
tutorial.

## Don't Sleep

> When waiting for a resource to become available or an operation to complete do not use fixed wait time,

Reviewer: '...do not use fixed wait times  and/or sleep statements...'

> instead you should create a loop and keep pooling until the operation doesn't complete.

Reviewer 2: ```suggestion
instead you should create a loop and check within that loop if the resource is available or operation has completed.
```

Author: I tweaked it a little bit agai

## Independent

> A test must always be independent of other tests, which mean:

Reviewer 2: ```suggestion
A test must always be independent of other tests, which means:
```

> - it should not rely on other tests changes so that it would be possible to run it independently

Reviewer: '...It should not rely on the successful execution of other tests...'

> - it should not rely on other tests changes so that it would be possible to run it independently

Reviewer 2: ```suggestion
- it should not rely on the successful execution of other tests so the test can be run independently
```

Author: @damienomurchu I don't get why removing the `so that`, I would like to emphasize that there is a reason why we do it and a consequence if you don't do it and I think that using `so that` will strength the sentence. I would like just o understand why removing it?

> - it should clean up after himself so that it would be possible to run it multiple time

Reviewer 2: ```suggestion
- it should clean up after itself so it is possible to run the test concurrently or multiple times
```

> - it should not rely on a static order when querying data because multiple tests may run in parallel

Reviewer 2: I think we should reword this somehow so its a bit clearer

Author: This is actually more an example of what not to do, maybe we can just remove this

> - it should not delete or modify other tests data because multiple tests may run in parallel

Reviewer 2: ```suggestion
- it should not delete or modify resources used by other tests so tests may run in parallel
```

> - it should not destruct any resource because other tests may rely on them

Reviewer 2: I think this point is redundant as its covered by the previous point

## Secrets

> Do not commit or log any sensitive data or secrets, and always double check for it because it can happen unconsciously

Reviewer 2: ```suggestion
Ensure secret or sensitive data is not included in commits or log output. 
```

## Logging

Do not be afraid of logging, if the test pass we will ignore the logs but if the test fail logs will help

> understand why the test has failed, especially if the tests fail during an unexpected step.

Reviewer 2: ```
Log statements should be used to add meaningful test execution information at regular points in the testcase. This is important so log output is useful and unnecessary additional, lengthy test runs to debug failures can be avoided. 
```

Reviewer 2: @b1zzu I think we should mention the use of logging levels here too

Author: Using the standard log testing library for go is not possible to specify a log level, but instead of introducing a new library for the moment using the standard testing library is already an improvement over the fmt.Print because it will print also the timestamp and the line number
